# Logistic Regression for Binary Classification on mnist dataset
                    

In [162]:
# importing packages

import numpy as np
from scipy.io import loadmat
from sklearn import metrics


In [163]:
# loading the training datasets

x_traindata = loadmat("mnistTrainImages.mat")
y_trainlabels = loadmat("mnistTrainLabels.mat")

# Extracting the relevant information from the datasets

x_traindata = x_traindata["trainData"]
y_trainlabels = y_trainlabels["trainLabels"]

print x_traindata.shape
print y_trainlabels.shape

(60000, 784)
(60000, 1)


In [164]:
# loading the test datasets

x_testdata = loadmat("mnistTestImages.mat")
y_testlabels = loadmat("mnistTestLabels.mat")

# Extracting the relevant information from the datase

x_testdata = x_testdata["testData"]
y_testlabels = y_testlabels["testLabels"]

#print x_testdata.shape
#print type(y_testlabels)

In [165]:
# converting the list of lists to a numpy array

y_trainlabels = [i[0] for i in y_trainlabels]
y_trainlabels = np.array(y_trainlabels)
print y_trainlabels.shape

y_testlabels = [i[0] for i in y_testlabels]
y_testlabels = np.array(y_testlabels)
print y_testlabels.shape


(60000,)
(10000,)


In [166]:
# Inserting a column of ones for both training and test data

x_traindata = np.insert(x_traindata,0,1,axis=1)
x_testdata = np.insert(x_testdata,0,1,axis=1)

print x_traindata.shape
#print x_testdata.shape

(60000, 785)


In [167]:
# Initializing the parameters

theta = np.ones(x_traindata[0].shape)
oldtheta = np.inf
alpha = 0.05
epsilon = 15
#print theta
print oldtheta
print theta.shape

inf
(785,)


In [168]:
def logit(x,theta_calc):
    return 1/(1+(np.exp(-(np.matmul(theta_calc,np.transpose(x))))))


In [169]:
def calc_loss_fun(x,theta_calc,y):
    return np.matmul(np.transpose(x),(logit(x,theta_calc)-y))

In [170]:
def Gradient_descent(learn_rate,tolerance,theta_calc,old_theta_calc,x,y):
    iteration = 0
    while(np.linalg.norm(theta_calc - old_theta_calc, ord=2) > tolerance):
        iteration = iteration+1
        old_theta_calc = theta_calc
        theta_calc = old_theta_calc - learn_rate*calc_loss_fun(x,theta_calc,y)
    return (theta_calc,iteration)
        

In [171]:
(theta,iteration) = Gradient_descent(alpha,epsilon,theta,oldtheta,x_traindata,y_trainlabels)
print theta.shape
print "\n Iteration = ", iteration

/home/dmacs/anaconda2/lib/python2.7/site-packages/ipykernel_launcher.py:2: RuntimeWarning: overflow encountered in exp
  


(785,)

 Iteration =  239


In [172]:
prediction = logit(x_testdata,theta)

/home/dmacs/anaconda2/lib/python2.7/site-packages/ipykernel_launcher.py:2: RuntimeWarning: overflow encountered in exp
  


In [173]:
# Assigning Labels 

cutoff = 0.5
for i in range(prediction.size):
    if prediction[i] >= cutoff:
        prediction[i] = 1
    else:
        prediction[i] = 0
print prediction.shape

(10000,)


In [175]:
# Calculating the accuracy

correct_prediction = 0
for i in range(y_testlabels.size):
    if prediction[i] == y_testlabels[i]:
        correct_prediction = correct_prediction +1
print "Accuracy is ", correct_prediction/float(y_testlabels.size)


Accuracy is  0.9907


 Observation - 


| Alpha(Learning_Rate) | Epsilon(Error_tollerance) | No._of_iteration | Accuracy |
| :--------: | :-------: | :---------: | :-------:|
| | | | |
| 0.1 | 30 | 239 | 99.07 |
| 0.1 | 20 | 387 | 99.12 |
| 0.1 | 15 | 1835 | 99.11 |
| | | | |
| 0.05 | 30 |107 | 99.02 |
| 0.05| 20 | 177 | 99.01 |
| 0.05 | 15 | 239 | 99.07 |
| | | | |
| 0.01 | 30 | 14 | 97.61 |
| 0.01 | 20 | 27 | 98.17 |
| 0.01 | 15 | 38 | 98.38 |

Though for learning rate 0.1 and error tollerance 15 we got a very good accuracy of 99.11 the number of iteration it took is almost 7 times more than the number of iteration it took for learning rate 0.05 with error tollerance 15 giving a close accuracy of 99.07.

It is also interesting to note that we got same accuracy i.e 99.07 for learning rate 0.1 and error tollerance 30 and for learning rate 0.05 and error tollerance 15. Both of these took same number of iteration i.e 239. Given a choice to pick between one of the two configuration, we must pick the one with least error tollerance in such case.

We also observe a very slight oscillation in accuracy when error tollerance is changed from 20 to 15 in case of learning rate 0.1 and when error rate was changed from 30 to 20 in case of learning rate 0.05. In both these cases the number of iteration increased but the accuracy decreased.